packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,360 @@
+import { Connection } from "@libp2p/interface-connection";
+import { PeerId } from "@libp2p/interface-peer-id";
+import { PeerInfo } from "@libp2p/interface-peer-info";
+import { Peer } from "@libp2p/interface-peer-store";
+import { IRelay, Tags } from "@waku/interfaces";
+import debug from "debug";
+import type { Libp2p } from "libp2p";
+
+import KeepAliveManager, { KeepAliveOptions } from "./keep_alive_manager.js";
+
+const DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED = 2;
+const DEFAULT_APPROX_TIME_BEFORE_BOOTSTRAP_FALLBACK_MS = 5000;
+const DEFAULT_MAX_DIAL_ATTEMPTS_FOR_PEER = 3;
+const DEFAULT_LOGGING_INTERVAL = 10_000;
+
+const log = debug("waku:connection-manager");
+
+export interface Libp2pComponents {
+  connectionManager: Libp2p["connectionManager"];
+  peerStore: Libp2p["peerStore"];
+  ping: Libp2p["ping"];
+  dial: Libp2p["dial"];
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+export interface Options {
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   * Is overridden by {@link maxDialAttemptsBeforeBootstrapFallback}
+   */
+  maxBootstrapPeersAllowed?: number;
+  /**
+   * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+   * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+   * This is only used when other discovery mechanisms haven't yield peers that are dialable
+   * This increases relative centralization of the network
+   */
+  maxTimeBeforeBootstrapFallbackMs?: number;
+}
+
+export class ConnectionManager extends KeepAliveManager {
```

Comment:
I am not convinced inheritance is the best way to go as conceptually the connection manager does not strike to me as a specialization of a keep alive manager. It also bloats the API.